src/main/main.c

```diff
@@ -411,60 +427,71 @@ static int set_kqueue_fds(struct re *re, re_sock_t fd, int flags)
 
 
 /**
- * Rebuild the file descriptor mapping table. This must be done whenever
+ * Rebuild the file descriptor mapping. This must be done whenever
  * the polling method is changed.
  */
-static int rebuild_fds(struct re *re)
+static bool rebuild_fd(struct le *he, void *arg)
 {
-	int i, err = 0;
+	int err = 0;
+	struct re *re = arg;
+	struct fhs *fhs = he->data;
 
-	DEBUG_INFO("rebuilding fds (nfds=%d)\n", re->nfds);
 
 	/* Update fd sets */
-	for (i=0; i<re->nfds; i++) {
-		if (!re->fhs[i].fh)
-			continue;
+	if (!fhs->fh)
+		return false;
 
-		switch (re->method) {
+	switch (re->method) {
 
 #ifdef HAVE_POLL
-		case METHOD_POLL:
-			err = set_poll_fds(re, i, re->fhs[i].flags);
-			break;
+	case METHOD_POLL:
+		err = set_poll_fds(re, fhs);
+		break;
 #endif
 #ifdef HAVE_EPOLL
-		case METHOD_EPOLL:
-			err = set_epoll_fds(re, i, re->fhs[i].flags);
-			break;
+	case METHOD_EPOLL:
+		err = set_epoll_fds(re, fhs);
+		break;
 #endif
 
 #ifdef HAVE_KQUEUE
-		case METHOD_KQUEUE:
-			err = set_kqueue_fds(re, i, re->fhs[i].flags);
-			break;
+	case METHOD_KQUEUE:
+		err = set_kqueue_fds(re, fhs);
+		break;
 #endif
 
-		default:
-			break;
-		}
+	default:
+		break;
+	}
 
-		if (err)
-			break;
+	if (err) {
+		DEBUG_WARNING("rebuild_fd: set fd error: %m\n", err);
+		return true;
 	}
 
-	return err;
+	return false;
 }
 
 
 static int poll_init(struct re *re)
 {
+	int err;
+
 	DEBUG_INFO("poll init (maxfds=%d)\n", re->maxfds);
 
 	if (!re->maxfds) {
 		DEBUG_WARNING("poll init: maxfds is 0\n");
 		return EINVAL;
 	}
 
+	if (!re->fhl) {
+		err = hash_alloc(&re->fhl, hash_valid_size(re->maxfds));
+		if (err)
+			return err;
+	}
+
+	list_flush(&re->fhs_delete);
```

Comment:
If this is called from `fd_close`, this causes the same problem of invalid `data.ptr` we discussed.